## Week 10 and 11 Assignment - DATASCI200 Introduction to Data Science Programming, UC Berkeley MIDS

Write code in this Jupyter Notebook to solve the following problems. Please upload this **Notebook** with your solutions to your GitHub repository and provide a link in the last question in gradescope. 

Assignment due date: 11:59PM PT the night before the Week 12 Live Session. Do **NOT** push/upload the data file. 

## Objectives

- Explore and glean insights from a real dataset using pandas
- Practice using pandas for exploratory analysis, information gathering, and discovery
- Practice cleaning data and answering questions

## General Guidelines:

- This is a **real** dataset and so it may contain errors and other pecularities to work through
- This dataset is ~218mb, which will take some time to load (and probably won't load in Google Sheets or Excel)
- If you make assumptions, annotate them in your responses
- While there is one code/markdown cell positioned after each question as a placeholder, some of your code/responses may require multiple cells
- Double-click the markdown cells that say for example **1a answer here:** to enter your written answers. If you need more cells for your written answers, make them markdown cells (rather than code cells)
- This homework assignment is not autograded because of the variety of responses one could give. 
  - Please upload this notebook to the autograder page and the TAs will manually grade it. 
  - Ensure that each cell is run and outputs your answer for ease of grading! 
  - Highly suggest to do a `restart & run all` before uploading your code to ensure everything runs and outputs correctly.
  - Answers without code (or code that runs) will be given 0 points.
- **This is meant to simulate real world data so you will have to do some external research to determine what some of the answers are!** 

## Dataset

You are to analyze campaign contributions to the 2016 U.S. presidential primary races made in California. Use the csv file located here: https://drive.google.com/file/d/1Lgg-PwXQ6TQLDowd6XyBxZw5g1NGWPjB/view?usp=sharing. You should download and save this file in the same folder as this notebook is stored.  This file originally came from the U.S. Federal Election Commission (https://www.fec.gov/).

**DO NOT PUSH THIS FILE TO YOUR GITHUB REPO!**

- Best practice is to not have DATA files in your code repo. As shown below, the default load is outside of the folder this notebook is in. If you change the folder where the file is stored please update the first cell!
- If you do accidentally push the file to your github repo - follow the directions here to fix it: https://docs.google.com/document/d/15Irgb5V5G7pKPWgAerH7FPMpKeQRunbNflaW-hR2hTA/edit?usp=sharing

Documentation for this data can be found here: https://drive.google.com/file/d/11o_SByceenv0NgNMstM-dxC1jL7I9fHL/view?usp=sharing

## Data Questions

You are working for a California state-wide election campaign. Your boss wants you to examine historic 2016 election contribution data to see what zipcodes are more supportive of fundraising for your candidate. 

Your boss asks you to filter out some of the records:
- Only use primary 2016 contribution data (more like how your race is).
- Concentrate on Bernie Sanders as a candidate (most a like your candidate)

The questions your boss wants answered is:
- Which zipcode (5-digit zipcode) had the highest count of contributions and the most dollar amount?
- What day(s) of the month do most people donate?

## Setup

Run the cell below as it will load the data into a pandas dataframe named `contrib`. Note that a custom date parser is defined to speed up loading. If Python were to guess the date format, it would take even longer to load.

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

# These commands below set some options for pandas and to have matplotlib show the charts in the notebook
pd.set_option('display.max_rows', 1000)
pd.options.display.float_format = '{:,.2f}'.format
#pd.set_option('display.max_rows', 100)
#pd.set_option('display.width', 500)

# Define a date parser to pass to read_csv
d = lambda x: datetime.strptime(x, '%d-%b-%y')

# Load the data
# We have this defaulted to the folder OUTSIDE of your repo - please change it as needed
contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)

# Note - for now, it is okay to ignore the warning about mixed types. 

C:\Users\jspen\AppData\Local\Temp\ipykernel_16732\1692960448.py:16: DtypeWarning: Columns (6,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  contrib = pd.read_csv('P00000001-CA.csv', index_col=False, parse_dates=['contb_receipt_dt'], date_parser=d)


***
## 1. Initial Data Checks (50 points)

First we will take a preliminary look at the data to check that it was loaded correctly and contains the info we need.

The questions to answer at the end of this section:
- Do we have the correct # of columns and rows. 
- Do the records contain data for the questions we want to answer 
- What columns are important? 
- What columns can be dropped?
- What are the data problems?

**1a.** Print the *shape* of the data. Does this match the expectation? (2 points)

In [2]:
# 1a YOUR CODE HERE
print('SHAPE:  ',contrib.shape)
#print(contrib.head())

SHAPE:   (1125659, 18)


- **1a answer here:** 
The dataframe's shape indicates there are 18 columns and 1125659 rows.  The number of columns matches what would be expected from the dataset documentation, and the number of rows is reasonable.

**1b.** Print a list of column names. Are all the columns included that are in the documentation? (2 points)

In [3]:
# 1b YOUR CODE HERE
print(contrib.columns)

Index(['cmte_id', 'cand_id', 'cand_nm', 'contbr_nm', 'contbr_city',
       'contbr_st', 'contbr_zip', 'contbr_employer', 'contbr_occupation',
       'contb_receipt_amt', 'contb_receipt_dt', 'receipt_desc', 'memo_cd',
       'memo_text', 'form_tp', 'file_num', 'tran_id', 'election_tp'],
      dtype='object')


- **1b answer here:** 
Yes, all columns specified in the documentation are found in the dataframe.

**1c** Print out the first five rows of the dataset. How do the columns `cand_id`, `cand_nm` and `contbr_st` look? (3 points)

In [4]:
# 1c YOUR CODE HERE
print(contrib.head(5))

     cmte_id    cand_id                  cand_nm          contbr_nm  \
0  C00575795  P00003392  Clinton, Hillary Rodham         AULL, ANNE   
1  C00575795  P00003392  Clinton, Hillary Rodham  CARROLL, MARYJEAN   
2  C00575795  P00003392  Clinton, Hillary Rodham   GANDARA, DESIREE   
3  C00577130  P60007168         Sanders, Bernard          LEE, ALAN   
4  C00577130  P60007168         Sanders, Bernard   LEONELLI, ODETTE   

     contbr_city contbr_st     contbr_zip            contbr_employer  \
0       LARKSPUR        CA 949,391,913.00                        NaN   
1        CAMBRIA        CA 934,284,638.00                        NaN   
2        FONTANA        CA 923,371,507.00                        NaN   
3      CAMARILLO        CA 930,111,214.00  AT&T GOVERNMENT SOLUTIONS   
4  REDONDO BEACH        CA 902,784,310.00   VERICOR ENTERPRISES INC.   

   contbr_occupation  contb_receipt_amt contb_receipt_dt receipt_desc memo_cd  \
0            RETIRED              50.00       2016-04-26   

- **1c answer here:** 
The cand_id column appears to be an identifier uniquely associated with the cand_nm, which shows the candidate's surname first.  Elements in both columns appear reasoanble.  The contbr_st appears to be a two-character string for California, which is expected.

**1d.** Print out the values for the column `election_tp`. In your own words, based on the documentation, what information does the `election_tp` variable contain? Do the values in the column match the documentation? (3 points)

In [5]:
# 1d YOUR CODE HERE
contrib['election_tp'].unique()

array(['P2016', 'G2016', nan, 'P2020'], dtype=object)

- **1d answer here:** 
According to the documentation, the values in the column election_tp are indicative of the election for which the contribution was made, and include an election prefix followed by the year.  The unique values for this field include P2016 and G2016, which are expected and correspond to the primary and general elections in 2016.  However we also observe values nan and P2020, indicating the dataset includes contributions with missing election type data, as well as contributions for the 2020 primary, both of which should be excluded from the final analysis.

**1e.** Print out the datatypes for all of the columns. What are the datatypes for the `contbr_zip`, `contb_receipt_amt`, `contb_receipt_dt`? (5 points)

In [6]:
# 1e YOUR CODE HERE
contrib.dtypes

cmte_id                      object
cand_id                      object
cand_nm                      object
contbr_nm                    object
contbr_city                  object
contbr_st                    object
contbr_zip                   object
contbr_employer              object
contbr_occupation            object
contb_receipt_amt           float64
contb_receipt_dt     datetime64[ns]
receipt_desc                 object
memo_cd                      object
memo_text                    object
form_tp                      object
file_num                      int64
tran_id                      object
election_tp                  object
dtype: object

- **1e answer here:** 
The contrib_zip column has an 'object' datatype which is unexpected as this field contain should contain zip codes, ideally as 5-digit integers.  The column contb_receipt_amt is float64 which is expected as this field contains values of currency, although float64 may be overkill and the column could optionally be converted to float32 to reduce memory requirements with no expected loss to data precision.  The column conb_receipt_dt is of datatype datetime64[ns] which is expected as this column the date of the contribution.

**1f.** What columns have the most non-nulls?  Would you recommend to drop any columns based on the number of nulls? (5 points)

In [7]:
# 1f YOUR CODE HERE
#print(np.sum(~contrib.isnull()))
print('\n',np.sum(contrib.isnull())/len(contrib))


 cmte_id             0.00
cand_id             0.00
cand_nm             0.00
contbr_nm           0.00
contbr_city         0.00
contbr_st           0.00
contbr_zip          0.00
contbr_employer     0.14
contbr_occupation   0.01
contb_receipt_amt   0.00
contb_receipt_dt    0.00
receipt_desc        0.99
memo_cd             0.87
memo_text           0.55
form_tp             0.00
file_num            0.00
tran_id             0.00
election_tp         0.00
dtype: float64


- **1f answer here:** 
Most of the columns have a large number of non-nulls.  The following columns have over 10% of values as null:  contb_employer, receipt_desc, memo_cd, memo_text.  Based on our objectives, all of these columns can be excluded from the analysis.  A few of the zipcodes are also missing, and as this column is of interest to us, these rows may be dropped during the analysis.

**1g.** A column we know that we want to use is the cand_nm column.  From the documentation each candidate is a unique candidate id also. Check data quality of `cand_id` column to see if it matches `cand_nm` column. Specifically check to ensure our targetted candidate 'Bernard Sanders' always has the same cand_id throughout. Any issues with `cand_nm` matching `cand_id`? (5 points)

In [8]:
# 1g YOUR CODE HERE
# return number of rows where the candidate id isn't the same candidate id as the first row, for a given candidate name
candidate_name = 'Sanders, Bernard'
candidate_id = contrib.loc[contrib[contrib['cand_nm']==candidate_name].index[0],'cand_id']
len(contrib[ (contrib['cand_nm']==candidate_name) & (contrib['cand_id']!=candidate_id) ])

0

- **1g answer here:** 
There are no Sanders records where there's a mismatch between the cand_nm and cand_id columns.

**1h.** Another area to check is to make sure all of the records are from California. Check the `contbr_st` column - are there any records outside of California based on `contbr_st`? (5 points)

In [9]:
# 1h YOUR CODE HERE
#len(contrib[ contrib['contbr_st'] != 'CA' ])
print(contrib['contbr_st'].unique())

['CA']


- **1h answer here:** 
There are no records of contributers from outside of California.

**1i.** The next column to check for the analysis is the `tran_id` column. This column could be the primary key so look for duplicates. How many duplicate entries are there? Any pattern for why are there duplicate entries? (5 points)

In [10]:
# 1i YOUR CODE HERE
print(contrib['tran_id'].is_unique)
replicated_tran_id = contrib[contrib.duplicated(subset='tran_id', keep=False)]
print(len(replicated_tran_id))
print(replicated_tran_id.head(20))
#print(contrib[contrib['tran_id']=='SA17.770299'])
#print(contrib[contrib['tran_id']=='SA17.771839'])
#print(contrib[contrib['tran_id']=='SA17.757766'])
#print(replicated_tran_id['tran_id'].describe())

False
6873
        cmte_id    cand_id                    cand_nm              contbr_nm  \
57    C00575795  P00003392    Clinton, Hillary Rodham        SANMUKH, BHAKTA   
90    C00575795  P00003392    Clinton, Hillary Rodham        ROBERT, MONKMAN   
91    C00575795  P00003392    Clinton, Hillary Rodham        DOROTHY, FIEGER   
174   C00574624  P60006111  Cruz, Rafael Edward 'Ted'        MORENO, KENNETH   
175   C00574624  P60006111  Cruz, Rafael Edward 'Ted'  NOWAKOWSKI, HENRY MR.   
197   C00575795  P00003392    Clinton, Hillary Rodham       JOHN, FROITZHEIM   
433   C00575795  P00003392    Clinton, Hillary Rodham          MAGID, DANIEL   
676   C00575795  P00003392    Clinton, Hillary Rodham        DIANE, HAVENNER   
740   C00575795  P00003392    Clinton, Hillary Rodham              PAUL, GEE   
743   C00575795  P00003392    Clinton, Hillary Rodham         PINKEL, MARITA   
886   C00575795  P00003392    Clinton, Hillary Rodham       THOMAS, HALLINAN   
1050  C00575795  P00003392   

- **1i answer here:** 
There are nearly 7000 records with replicated values of tran_id.  At first glance this is unexpected, but after more closely reviewing the documentation, it appears that the file_num in combination with the tran_id uniquely identify a given transaction.  Clearly the tran_id alone is not a suitable unique identifier for each row.

**1j.** Another column to check is the `contb_receipt_amt` that shows the donation amounts. How many negative donations are included? What do negative donations mean? Please show at least pull a few rows to look at the records with negative donations. Do these records match with the expectation of why a negative donation would happen? (5 points)

In [11]:
# 1j YOUR CODE HERE
print(len(contrib[ contrib['contb_receipt_amt'] < 0 ]))
#print(contrib[ contrib['contb_receipt_amt'] < 0 ].head(5))
negative_contributions = contrib.loc[ contrib['contb_receipt_amt'] < 0 ]
print(negative_contributions['memo_text'].unique())

11896
['REDESIGNATION TO GENERAL' nan '* HILLARY VICTORY FUND'
 'REATTRIBUTION TO SPOUSE' 'CHARGED BACK'
 'REDESIGNATION TO PRESIDENTIAL GENERAL' 'TRANSFER FROM RUBIO VICTORY'
 '2016 SENATE PRIMARY DONOR REDESIGNATION TO GENERAL'
 'REDESIGNATION FROM PRIMARY' 'US CITIZEN REDESIGNATION TO GENERAL'
 'REDESIGNATION TO CRUZ FOR SENATE' '* REATTRIBUTED TO JOHN CASTOR'
 'NSF/RETURNED CHECK' '* REATTRIBUTED TO FRANK DARABONT'
 'REFUND TO BE ISSUED; REDESIGNATION TO GENERAL'
 'REFUND TO BE ISSUED; CHARGED BACK'
 'REATTRIBUTION TO SPOUSE; SEE REDESIGNATION' 'CHECK RETURNED BY BANK'
 'OVERPAYMENT' '*BEST EFFORTS UPDATE'
 '* REATTRIBUTED TO CHRISTINE MARSHALL'
 '* REATTRIBUTED TO CAROL FITZMORRIS' '* REATTRIBUTED TO ERIC UHRANE'
 '* REATTRIBUTED TO YEU-WEI YEE' '* REATTRIBUTED TO NORAH STONE'
 '* REATTRIBUTED FROM KIM RIDDER' 'INSUFFICIENT FUNDS'
 'REDESIGNATION TO PRIMARY' 'REATTRIBUTION TO SPOUSE, TERRIE SCHULTZ.'
 '* HILLARY ACTION FUND' 'VOIDED 9/16/16' '2016 SENATE GENERAL'
 '* REATTRIBUTION

- **1j answer here:**
There are 11896 rows with negative donation amounts.  Negative transactions are attributable to several reasons, including redesignation from a primary campaign to a general campaign, refund to the contributor, and reattribution to another candidate or committee.

**1k.** One more column to look at is the date of donation column. Are there any dates outside of the primary period (defined as 1 Jan 2014 to 7 June 2016)? Are the dates well-formatted for our analysis? (5 points)

In [12]:
# 1k YOUR CODE HERE
print(contrib['contb_receipt_dt'].describe(datetime_is_numeric=True))
#print(contrib['contb_receipt_dt'].info())
#print(contrib['contb_receipt_dt'].tail(10))
primary_begin_date = pd.to_datetime('2014-01-01')
primary_end_date = pd.to_datetime('2016-06-07')
invalid_date_pct = len(contrib[ (contrib['contb_receipt_dt'] < primary_begin_date) | (contrib['contb_receipt_dt'] > primary_end_date) ]) / len(contrib) 
print(invalid_date_pct)

count                          1125659
mean     2016-05-03 12:17:40.562745856
min                2013-11-05 00:00:00
25%                2016-02-29 00:00:00
50%                2016-05-03 00:00:00
75%                2016-08-01 00:00:00
max                2016-10-19 00:00:00
Name: contb_receipt_dt, dtype: object
0.4009846676480177


- **1k answer here:**
Yes, about 40% of the rows correspond to donations made before or after the valid 2016 primary dates.  The dates do seem to be consistently formatted for analysis.

**1l.** Finally, answer the initial questions in the cells below (5 points)

**1l.1** Do we have the correct # of columns and rows.

- **1l.1 answer here:**
The original dataset has the correct number of rows and columns.  However we will need to filter the dataset appropriately prior to our analysis, since we are interested in a particular subset of the data.

**1l.2** Do the records contain data for the questions we want to answer?

- **1l.2 answer here:**
Yes, the dataset should be suitable to answer the questions.

**1l.3** What columns are important?

- **1l.3 answer here:** 
The most important columns are:  cand_nm, contbr_zip, contb_receipt_amt, contb_receipt_dt.  

**1l.4** What columns can be dropped?

- **1l.4 answer here:** 
We can immediately drop the columns with significant null values:  contb_employer, receipt_desc, memo_cd, memo_text.  For the final analysis, given its limited objectives, we can perform all of the data subsetting and eventually drop all but the 4 columns above.

**1l.5** What are the data problems?

- **1l.5 answer here:**
The zipcodes are not formatted as integers and contain null values, so that column will need to be treated accordingly since one of our primary questions relates to zip code.  We will also need to create a new column that extracts the day of the month from the contribution date, as that will be required to answer the other question.

**1l.6** List any assumptions so far:

- **1l.6 answer here:**
We will assume the form_tp, file_num, and tran_id data are compliant with regulatory requirements, and requires no additional validations.
We will assume all contributions, regardless of the details (eg. as captured in the memo_text field) are valid for our analysis.

***
## 2. Data filtering and data quality fixes (30 points)

Now that we have a basic understanding of the data, let's filter out the records we don't need and fix the data.

**2a.** From the dataset filter out (remove) any election_tp not in the primary election. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [13]:
# 2a YOUR CODE HERE
contrib_subset = contrib[ contrib['election_tp'] == 'P2016' ].copy()
#print(contrib_subset.head())
print(contrib_subset.shape)

(810481, 18)


**2b.** From the dataset filter out (remove) any candidate that is not Bernie Sanders. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [14]:
# 2b YOUR CODE HERE
#print(contrib_subset['cand_nm'].head(100))
contrib_subset = contrib_subset [ contrib_subset['cand_nm'] == 'Sanders, Bernard' ].copy().reset_index(drop=True)
#print(contrib_subset.head())
print(contrib_subset.shape)

(407171, 18)


**2c.** The `contbr_zip` column is not formatted well for our analysis. Make a new zipcode column that is the five-digit zipcodes. Filter out any records outside of California based on the zipcode. Print/show the shape of the dataframe after the filtering is complete. (10 points).

- You will have to research what the valid 5-digit zipcodes for California are!

In [15]:
# 2c YOUR CODE HERE
# according to wikipedia, California is assigned ZIP codes ranging from 900-961
# ref: https://en.wikipedia.org/wiki/ZIP_Code
contrib_subset['contbr_zip'] = contrib_subset['contbr_zip'].astype(str).str.replace(',','').replace('.','').str[:5]
contrib_subset = contrib_subset[ contrib_subset['contbr_zip'].map(len) == 5]
contrib_subset['contbr_zip'] = contrib_subset['contbr_zip'].apply(pd.to_numeric, errors='coerce').fillna(0.0).astype(int)
contrib_subset = contrib_subset [ (contrib_subset['contbr_zip'] >= 90000) | (contrib_subset['contbr_zip'] <= 96199) ].reset_index(drop=True)
print(contrib_subset.shape)

(407167, 18)


**2d.** The receipt amount column has negative donations. After talking with your team, a decision was made that the best course of action is to remove these negative values so that the donation count and amount is more accurate. Print/show the shape of the dataframe after the filtering is complete. (5 points)

In [16]:
# 2d YOUR CODE HERE
contrib_subset = contrib_subset [ contrib_subset['contb_receipt_amt'] > 0 ].reset_index(drop=True)
print(contrib_subset.shape)

(404078, 18)


**2e.** From the dataset drop any columns that won't be used in the analysis. Print/show the shape of the dataframe after the dropping is complete. What columns did you drop and why? (5 points)

In [17]:
# 2e YOUR CODE HERE
relevant_cols = ['cand_nm','contbr_zip','contb_receipt_amt','contb_receipt_dt']
contrib_subset = contrib_subset[ contrib_subset.columns.intersection(relevant_cols) ]
print(contrib_subset.head())
print(contrib_subset.shape)

            cand_nm  contbr_zip  contb_receipt_amt contb_receipt_dt
0  Sanders, Bernard       93011              40.00       2016-03-04
1  Sanders, Bernard       90278              35.00       2016-03-05
2  Sanders, Bernard       90278             100.00       2016-03-06
3  Sanders, Bernard       92084              25.00       2016-03-04
4  Sanders, Bernard       92683              10.00       2016-03-05
(404078, 4)


- **2e answer here:**

**2f.** List any assumptions that you made up to this point:

- **2f answer here:**
We will assume the form_tp, file_num, and tran_id data are compliant with regulatory requirements, and requires no additional validations.
We will assume all contributions, regardless of the details (eg. as captured in the memo_text field) are valid for our analysis.
We will not consider any rows for which the election_tp is null.
We will not consider any rows for which the zip code is null or improperly formatted.
We will not consider any rows corresponding to negative contributions.

***
## 3. Answering the questions (20 points)

Now that the data is cleaned and filterd - let's answer the two questions from your boss!

**3a.** Which zipcode had the highest count of contributions and the most dollar amount? (10 points)

In [18]:
# 3a YOUR CODE HERE
print(contrib_subset['contbr_zip'].value_counts())
zip_highest_count = contrib_subset['contbr_zip'].mode()
print('zipcode with the highest count of contributions is:  ',zip_highest_count.values[0])

print(contrib_subset.groupby(['contbr_zip'])['contb_receipt_amt'].sum().sort_values(ascending=False))
#contrib_subset.loc[contrib_subset['contbr_zip']==94110].contb_receipt_amt.sum()
zip_highest_contb = contrib_subset.groupby(['contbr_zip'])['contb_receipt_amt'].sum().sort_values(ascending=False)
print('zipcode with the highest overall dollar amount of contributions is:  ',zip_highest_contb.index[0])

94110    4046
94114    3043
94117    2884
95060    2849
94611    2516
         ... 
96023       1
92059       1
93285       1
91614       1
92334       1
Name: contbr_zip, Length: 1840, dtype: int64
zipcode with the highest count of contributions is:   94110
contbr_zip
94110   294,061.13
94114   208,602.37
94117   201,545.43
94611   150,232.12
95060   145,414.63
           ...    
95703        15.00
95233        15.00
95701        15.00
93483        10.00
92365        10.00
Name: contb_receipt_amt, Length: 1840, dtype: float64
zipcode with the highest overall dollar amount of contributions is:   94110


- **3a answer here:** 
The zipcode with both the highest number of contributions and the largest overall contribution amount in dollars is 94110.

**3b.** What day(s) of the month do most people donate? (10 points)

In [19]:
# 3b YOUR CODE HERE
contrib_subset['contb_receipt_day_of_month'] = contrib_subset['contb_receipt_dt'].astype(str).str.slice(8,10).astype(int)
print(contrib_subset['contb_receipt_day_of_month'].value_counts().head())
day_of_month_highest_count = contrib_subset['contb_receipt_day_of_month'].mode()
print('zipcode with the highest count of contributions is:  ',day_of_month_highest_count.values[0])

29    23511
30    22903
31    19488
14    17793
8     16774
Name: contb_receipt_day_of_month, dtype: int64
zipcode with the highest count of contributions is:   29


- **3b answer here:** 
The 29th of the month seems to have the highest number of contributions.  There is a clear pattern that contributions are highest in the last few days of the month.

## If you have feedback for this homework, please submit it using the link below:

http://goo.gl/forms/74yCiQTf6k